[Data Views] User patterns: Use excerpt as description

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Enable descriptions for user created patterns by using the excerpt field.
Partial for https://github.com/WordPress/gutenberg/issues/55244

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
User created patterns did not have an option to add descriptions. Text only descriptions are important for users who are unable to see or interpret the content in the pattern preview.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This PR:
- Enables excerpts on the `wp_block` post type (This PHP change will eventually need a backport to `create_initial_post_types()` in wp-includes / post.php).
- On the Site Editor Pattern pages, use the excerpt as the description for user created patterns by:
   - Fetching the raw excerpt in `convertPatternPostToItem` and assigning it to the description.
   - Updating the condition that prints the description, to also print it on user created patterns.

## Todo:
The PR does not populate the excerpt field with the description when a theme or plugin pattern is duplicated into a user created pattern.

## Testing Instructions
The excerpt field can be tested both with a classic theme and block theme, when creating a new pattern or editing an existing pattern. The field should show in the "Pattern" sidebar panel, below the categories.

The Site Editor Patterns page can only be tested with a block theme:
Activate a block theme.
Create a new pattern. (It doesn't matter if the pattern is synced or not)
Add an excerpt to the pattern, save.
Go to Editor > Patterns, select "My patterns".
If you have not already done so, activate your screen reader.
Navigate to the pattern that you just created: When you hear that the title of your pattern is announced, it should be followed by the excerpt.

You can also confirm that the  excerpt text is output in the markup. It should be printed in a div inside another div with the class `dataviews-view-grid__media`

## Screenshots or screencast <!-- if applicable -->

<img width="279" alt="pattern-excerpt" src="https://github.com/WordPress/gutenberg/assets/7422055/4fff0adb-648a-4f0f-970e-2e80915df6c7">

carolinan committed:
[Data Views] User patterns: Use excerpt as description

Enable excerpts on the w_block post type.

On the Site Editor Pattern pages, use the excerpt as the description for user created patterns.

carolinan committed:
CS fix: Remove duplicate empty line

talldan commented:
This would be a great change to get in, thanks for working on it.

It's working well for me in testing. It might also be good to announce the description in the inserter when a pattern is focused:
![Screenshot 2024-04-08 at 12 50 28 pm](https://github.com/WordPress/gutenberg/assets/677833/c75b25dd-4e97-42a9-be9a-8d5ea44d2708)

And there are a couple of places it might be good to show it visually too, in the details sidebar in the site editor:
![Screenshot 2024-04-08 at 12 46 16 pm](https://github.com/WordPress/gutenberg/assets/677833/18091fca-d37f-4cff-ab44-fdbc8a5b6cb6)

And also in the block card in the inspector (though I think this one is low priority, and I'm not sure how easy it is):
![Screenshot 2024-04-08 at 12 48 51 pm](https://github.com/WordPress/gutenberg/assets/677833/8bbc1f77-e677-4079-b6f4-b1f3a5f9046d)

My feeling is that it might be nice to change the 'Excerpt' / 'Write an excerpt' labelling to 'Description', as I imagine it might be called that elsewhere (like in the table/grid view for patterns), but that could be something that's open to opinion, so would be good to discuss.

carolinan commented:
I agree, I did not explore those cases because I wanted to keep the initial change small.

carolinan committed:
Merge branch 'trunk' into update/pattern-pages-excerpt-as-description

carolinan commented:
I don't think it is ideal that the description is announced together with both the block name and the editing instructions.
On macOS, VoiceOver reads it our like this:
Pattern name Press Enter to Edit, or Delete to delete the pattern. Description, button, group

I found this confusing because it is not clear when the name, instructions, and description ends.

carolinan committed:
Place the description before the instructions.

carolinan commented:
I changed the position of the description and the instructions and added an em dash. But there is still no pause between the pattern name and the description:

Pattern name description —Press Enter to Edit, or Delete to delete the pattern., button, group

talldan commented:
> Press Enter to Edit, or Delete to delete the pattern

Oh, I'm surprised that text is still there. I think it can be removed now as the pattern grid has been completely replaced. It was only present before because there was no way to use the keyboard to reach the delete button. These days you can tab through every focusable element in the list.

In my testing, the delete key also no longer works, so there's that too.

carolinan commented:
On the Editor > Patterns screen, patterns that are **not created by users** does not have their description announced to screen reader users, nor visibly through a tooltip or similar.

The description should be announced in the same location on all types of pattern, because inconsistencies will make the pattern screens more difficult to use.

Related to **but not caused by** https://github.com/WordPress/gutenberg/pull/60653

I could open a separate issue, but since the description really should be in one consistent location, it could be fixed in this PR if we can agree on a better way to announce it. This PR would be blocked until that is solved...

carolinan committed:
Remove the incorrect instructions from ariaDescribedBy

- Removes the instructions from `ariaDescribedBy`, since they were not working and are not needed anymore.

- Because only one piece of content is added to `ariaDescribedBy`, the `ariaDescriptions` array and the `.map()` are not needed anymore, and are removed.

carolinan commented:
I don't know what has already been tried. Could we not use the button for all pattern types and parts and make sure the preview is always focusable, but disable the button with `aria-disabled` when the pattern can't be edited?

carolinan committed:
Try using the button and the description for all pattern types.

- Remove the condition that hides the item description on non user patterns.
- Use the button for all pattern types and for parts. Disable the button for patterns that are not editable.
- Add a title attribute on the non user patterns, to try to avoid confusion for sighted users that clicks the disabled button.

ntsekouras commented:
Thanks for the PR! I haven't checked the code yet, but I thought I'd share some first thoughts.

>And there are a couple of places it might be good to show it visually too, in the details sidebar in the site editor:

We are working towards removing the details sidebar, so we shouldn't be adding anything new there. The goal is to move things in the inspector controls sidebar and there are designs for excerpt too.



>I don't know what has already been tried. Could we not use the button for all pattern types and parts and make sure the preview is always focusable, but disable the button with aria-disabled when the pattern can't be edited?

I think that's fine if the buttons is disabled properly. I'd do it separately though the style changes it needs are a lot (that's my impression).

>Oh, I'm surprised that text is still there. I think it can be removed now as the pattern grid has been completely replaced. It was only present before because there was no way to use the keyboard to reach the delete button

I didn't know about that. We could also handle separately.

ntsekouras commented:
>@ntsekouras, do you have a link to an issue or the design file referenced here?

Designs about the move of excerpt are here: https://github.com/WordPress/gutenberg/issues/59689

carolinan committed:
Merge branch 'trunk' into update/pattern-pages-excerpt-as-description

carolinan commented:
When the pattern is not editable:
- The button is focusable with the keyboard and clickable, but there is no visual response for sighted users, all they see is that nothing happens.
- The button is clickable with the mouse. I added the title attribute so that when you hover over the button, there is a visible title text that says "This pattern can not be edited".

That different things happen when you click on different types of patterns is not a great experience.

Things to consider:
- If we are making theme patterns editable just like theme templates and parts are editable, then this becomes a non-issue, because the button wont be disabled.

-Or, if we need a solution to the description **_before_** the patterns are all editable, consider not using the button on the non-editable patterns, and instead move the description for all patterns from the button and make it visible to all users as part of the interface.

ntsekouras commented on lib/compat/wordpress-6.6/blocks.php:
```diff
@@ -0,0 +1,12 @@
+ * Adds support for excerpt to the wp_block post type.
+ */
+
+add_post_type_support( 'wp_block', 'excerpt' );
```
Should we rename this to `post.php` as this is the name in core that creates the post types?  Also maybe call in `init` action?

ntsekouras commented on packages/edit-site/src/components/page-patterns/index.js:
```diff
@@ -119,16 +119,25 @@ const SYNC_FILTERS = [
 			aria-describedby={ ariaDescribedBy }
+			aria-disabled={ item.type === PATTERN_TYPES.theme }
+			title={
+				item.type === PATTERN_TYPES.theme
```
Do we need to add this if it's disabled and we properly set the aria-disabled and onClick is noop? Wouldn't it work by just changing the cursor for visual response?

carolinan replied:
I removed the title attribute and changed the cursor when the button is disabled.

ntsekouras commented on packages/edit-site/src/components/page-patterns/index.js:
```diff
@@ -119,16 +119,25 @@ const SYNC_FILTERS = [
+				} else {
+					onClick( event );
+				}
+			} }
```
Nit:

```suggestion
			onClick={ item.type !== PATTERN_TYPES.theme ? onClick : undefined }
```

ntsekouras reviewed (commented):
I've left a couple of small comments, but this looks good to me. With this PR and when the excerpt update [PR](https://github.com/WordPress/gutenberg/pull/60894) lands, we'll also have the proper labelling (`description`).

A follow up could be to update the patterns list (Data Views) to add this field.

carolinan committed:
Remove title attribute and replace cursor style

Rename blocks.php to post.php.
Remove the title attribute from the button.
Reset the cursor style to the default when the button is disabled.

carolinan committed:
CS fix: Add empty line at end of file

ntsekouras reviewed (approved):
LGTM, thanks!

An issue that I observed is the display of the preview in `table` view is different between the patterns that are editable and the ones that are not. This is on trunk too, so we could land this and follow up with that. I'll cc @jameskoster if he can come up with some quick css to handle that.

jameskoster committed:
Fix pattern preview dimensions

jameskoster commented:
Good spot @ntsekouras, I pushed a fix.

jameskoster committed:
Update dimensions to match templates

carolinan commented:
Trac ticket: https://core.trac.wordpress.org/ticket/61250
